File: backend/src/vis_platform_backend/contracts/workspace_sessions.py
"""Workspace conversations: separate threads of assistant turns within one project."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from .assistant_turns import AssistantTurnLinks, AssistantTurnSnapshot
from .common import SCHEMA_VERSION, Identifier, StrictModel
from .plot_runs import PlotRunSnapshot
from .questions import PlannerAnswerRequest, PlannerQuestions


class CreateWorkspaceSession(StrictModel):
    schema_version: Literal["1.0"] = SCHEMA_VERSION
    request_id: Identifier
    title: str = Field(min_length=1, max_length=200)


class WorkspaceSession(StrictModel):
    schema_version: Literal["1.0"] = SCHEMA_VERSION
    session_id: str
    project_id: str
    title: str
    created_at: datetime
    updated_at: datetime = Field(description="When the conversation last received a message.")


class WorkspaceSessionList(StrictModel):
    sessions: list[WorkspaceSession]
    total: int
    offset: int


class AnsweredQuestions(StrictModel):
    question: PlannerQuestions
    answer: PlannerAnswerRequest


class WorkspaceSessionTurn(StrictModel):
    turn: AssistantTurnSnapshot
    links: AssistantTurnLinks
    answers: list[AnsweredQuestions] = Field(
        description="Questions the assistant asked during this turn, with the answers given."
    )
    run: PlotRunSnapshot | None = Field(description="The plot or analysis run the turn started.")


class WorkspaceSessionDocument(WorkspaceSession):
    turns: list[WorkspaceSessionTurn] = Field(description="Oldest first.")
    figure: PlotRunSnapshot | None = Field(
        description=(
            "The current version of the latest figure this conversation created, "
            "including later parameter changes and restores."
        )
    )

File: frontend/src/api/schemas/workspaceSessions.ts
import { z } from "zod";
import { plannerQuestionsSchema } from "./planner";
import {
  assistantLinksSchema,
  assistantTurnSnapshotSchema,
  plotRunSnapshotSchema,
} from "./plotRun";

export const workspaceSessionSchema = z
  .object({
    schema_version: z.literal("1.0"),
    session_id: z.string(),
    project_id: z.string(),
    title: z.string(),
    created_at: z.string(),
    updated_at: z.string(),
  })
  .strict();
export const workspaceSessionListSchema = z
  .object({
    sessions: z.array(workspaceSessionSchema),
    total: z.number().int(),
    offset: z.number().int(),
  })
  .strict();
const answeredQuestionsSchema = z
  .object({
    question: plannerQuestionsSchema,
    answer: z
      .object({
        project_id: z.string(),
        interaction_id: z.string(),
        answers: z.array(
          z
            .object({
              question_id: z.string(),
              choice_ids: z.array(z.string()),
              free_text: z.string().nullable().optional(),
            })
            .strict(),
        ),
      })
      .strict(),
  })
  .strict();
export const workspaceSessionDocumentSchema = workspaceSessionSchema
  .extend({
    turns: z.array(
      z
        .object({
          turn: assistantTurnSnapshotSchema,
          links: assistantLinksSchema,
          answers: z.array(answeredQuestionsSchema),
          run: plotRunSnapshotSchema.nullable(),
        })
        .strict(),
    ),
    figure: plotRunSnapshotSchema.nullable(),
  })
  .strict();

export type WorkspaceSession = z.infer<typeof workspaceSessionSchema>;
export type WorkspaceSessionDocument = z.infer<
  typeof workspaceSessionDocumentSchema
>;
